File: src/_cpp/TowerDefenseBase/scenes/ScoreLayer.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "ml/SmartScene.h"
#include "ScoreLayer.h"
#include "ScoreCounter.h"
#include "ShopLayer.h"
#include "configuration.h"
NS_CC_BEGIN

ScoreLayer::ScoreLayer()
{
	ScoreCounter::shared().observer( kScoreTime ).add( _ID, std::bind( &ScoreLayer::change_time, this, std::placeholders::_1 ) );
	ScoreCounter::shared().observer( kScoreFuel ).add( _ID, std::bind( &ScoreLayer::change_fuel, this, std::placeholders::_1 ) );
	ScoreCounter::shared().observer( kScoreCrystals ).add( _ID, std::bind( &ScoreLayer::change_real, this, std::placeholders::_1 ) );
	ScoreCounter::shared().observer( kScoreStars ).add( _ID, std::bind( &ScoreLayer::change_star, this, std::placeholders::_1 ) );
}

ScoreLayer::~ScoreLayer()
{
	ScoreCounter::shared().observer( kScoreTime ).remove( _ID );
	ScoreCounter::shared().observer( kScoreFuel ).remove( _ID );
	ScoreCounter::shared().observer( kScoreCrystals ).remove( _ID );
	ScoreCounter::shared().observer( kScoreStars ).remove( _ID );
}

bool ScoreLayer::init()
{
	do
	{
		CC_BREAK_IF( !Layer::init() );
		CC_BREAK_IF( !NodeExt::init() );

        if (k::configuration::useFuel) {
            NodeExt::load( "ini", "scorelayer_fuel.xml" );
        } else {
            NodeExt::load( "ini", "scorelayer.xml" );
        }

		_gold = getChildByName<Label*>( "valuegold" );
		_fuel = getChildByName<Label*>( "valuefuel" );
		_time = getChildByName<Label*>( "valuetime" );
		_star = getChildByName<Label*>( "valuestar" );
		_shop = dynamic_cast<MenuItem*>(getNodeByPath( this, "menu/shop" ));
		CC_BREAK_IF( !_gold );
		CC_BREAK_IF( !_fuel );
		CC_BREAK_IF( !_time );
		CC_BREAK_IF( !_shop );
		CC_BREAK_IF( !_star );

		change_fuel( ScoreCounter::shared().getMoney( kScoreFuel ) );
		change_time( ScoreCounter::shared().getMoney( kScoreTime ) );
		change_real( ScoreCounter::shared().getMoney( kScoreCrystals ) );
		change_star( ScoreCounter::shared().getMoney( kScoreStars ) );
		if( _shop )
		{
			_shop->setCallback( std::bind( &ScoreLayer::cb_shop, this, std::placeholders::_1 ) );
			if( k::configuration::useInapps == false )
				_shop->setVisible( false );
		}

		if( k::configuration::useFuel == false )
		{
			_fuel->setVisible( false );
			_time->setVisible( false );
		}

		return true;
	}
	while( false );
	return false;
}

void ScoreLayer::change_time( int score )
{
	int min = int( score ) / 60;
	int sec = int( score ) - min * 60;

	std::stringstream ss;
	if( min != 0 ) ss << min << "m ";
	if( sec != 0 ) ss << sec << "s";
	_time->setString( ss.str() );
}

void ScoreLayer::change_fuel( int score )
{
	bool vis = score < ScoreByTime::shared().max_fuel();
	_fuel->setString( intToStr( score ) );
	_time->setVisible( vis );

	if( k::configuration::useFuel == false )
	{
		_fuel->setVisible( false );
		_time->setVisible( false );
	}
}

void ScoreLayer::change_real( int score )
{
	_gold->setString( intToStr( score ) );
}

void ScoreLayer::change_star( int score )
{
	_star->setString( intToStr( score ) );
}

void ScoreLayer::cb_shop( Ref*sender )
{
#if PC != 1
	SmartScene * scene = dynamic_cast<SmartScene*>(getScene());
	if( scene->getChildByName( "shop" ) ) return;

	auto shop = ShopLayer::create( k::configuration::useFreeFuel, true, false, false );
	if( shop )
	{
		scene->pushLayer( shop, true );
	}
#endif
}




NS_CC_END

File: src/_eu/scenes/ScoreLayer.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/
/**TESTED**/
//Define namespace
var EU = EU || {};


EU.ScoreLayer = cc.Scene.extend({
    
    /** @type {Label} */ _gold : null,
    /** @type {Label} */ _fuel : null,
    /** @type {Label} */ _time : null,
    /** @type {Label} */ _star : null,
    /** @type {MenuItem} */ _shop : null,


    ctor: function()
    {
        this._super();
        this.init();
        this.initExt();
    },

    onEnter: function()
    {
        this._super();
        EU.ScoreCounter.observer( EU.kScoreTime ).add( this.__instanceId, this.change_time, this );
        EU.ScoreCounter.observer( EU.kScoreFuel ).add( this.__instanceId, this.change_fuel, this );
        EU.ScoreCounter.observer( EU.kScoreCrystals ).add( this.__instanceId, this.change_real, this );
        EU.ScoreCounter.observer( EU.kScoreStars ).add( this.__instanceId, this.change_star, this );
        this.updateScore();
    },
    onExit: function()
    {
        this._super();
        EU.ScoreCounter.observer( EU.kScoreTime ).remove( this.__instanceId );
        EU.ScoreCounter.observer( EU.kScoreFuel ).remove( this.__instanceId );
        EU.ScoreCounter.observer( EU.kScoreCrystals ).remove( this.__instanceId );
        EU.ScoreCounter.observer( EU.kScoreStars ).remove( this.__instanceId );
    },

    init: function()
    {
        if (EU.k.useFuel) {
            this.load_str( "ini/scorelayer_fuel.xml" );
        } else {
            this.load_str( "ini/scorelayer.xml" );
        }

        this._gold = this.getChildByName( "valuegold" );
        this._fuel = this.getChildByName( "valuefuel" );
        this._time = this.getChildByName( "valuetime" );
        this._star = this.getChildByName( "valuestar" );
        this._shop = EU.Common.getNodeByPath( this, "menu/shop" );
        if ( !this._gold ) return false;
        if ( !this._fuel ) return false;
        if ( !this._time ) return false;
        if ( !this._shop ) return false;
        if ( !this._star ) return false;

        this.change_fuel( EU.ScoreCounter.getMoney( EU.kScoreFuel ) );
        this.change_time( EU.ScoreCounter.getMoney( EU.kScoreTime ) );
        this.change_real( EU.ScoreCounter.getMoney( EU.kScoreCrystals ) );
        this.change_star( EU.ScoreCounter.getMoney( EU.kScoreStars ) );
        if( this._shop )
        {
            this._shop.setCallback( this.cb_shop, this );
            if( EU.k.useInapps == false )
                this._shop.setVisible( false );
        }

        if( EU.k.useFuel == false )
        {
            this._fuel.setVisible( false );
            this._time.setVisible( false );
        }

        return true;
    },

    updateScore: function(){
        this.change_time( EU.ScoreCounter.getMoney(EU.kScoreTime) );
        this.change_fuel( EU.ScoreCounter.getMoney(EU.kScoreFuel) );
        this.change_real( EU.ScoreCounter.getMoney(EU.kScoreCrystals) );
    },

    change_time : function(score )
    {
        var min = parseInt( score ) / 60;
        var sec = parseInt( score ) - min * 60;

        var ss;
        if( min != 0 ) ss = ss + min + "m ";
        if( sec != 0 ) ss = ss + sec + "s";
        /** Preven jsb fault */
        ss = ss || "";
        this._time.setString( ss );
    },

    change_fuel : function (score )
    {
        var vis = score < EU.ScoreByTime.max_fuel();
        this._fuel.setString( score);
        this._time.setVisible( vis );

        if( EU.k.useFuel == false )
        {
            this._fuel.setVisible( false );
            this._time.setVisible( false );
        }
    },

    change_real : function (score )
    {
        this._gold.setString( ( score ) );
    },

    change_star : function (score )
    {
        this._star.setString( ( score ) );
    },

    cb_shop: function(sender) {
        var scene = EU.Common.getSceneOfNode(this);
        if (scene.getChildByName("shop")) return;

        var shop = EU.ShopLayer.create(EU.k.useFreeFuel, true, false, false);
        if (shop) {
            scene.pushLayer(shop, true);
        }
    }

});

EU.NodeExt.call(EU.ScoreLayer.prototype);
